### PR TITLE
new database.yml format

### PR DESCRIPTION
**Pull request by arthurnn**
### Summary

New database.yml format. This is how it looks:
```ruby
default: &default
  adapter: sqlite3
  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
  timeout: 5000

development:
  primary:
    <<: *default
    database: db/development.sqlite3
  readonly:
    <<: *default
    database: db/readonly.sqlite3
```
We are moving from a two levels config to a three levels.

**This is backward compatible, apps using the default, old, 2 levels database.yml will still work.** 

### Advantages of this change
- Better structure on the database.yml file when dealing with multiples databases.
- One environment can have multiple configs.
- Having all the configs under an environment key, enables us to create/drop multiple databases on the create/drop tasks.
- Removes the necessity of adding the `Rails.env` when calling `establish_connection` for a non standard connection: i.e: `establish_connection "#{Rails.env}_readonly`

### Implementation details
The config file might have 3 levels, however, ActiveRecord now has a `local_configurations` option, which is scoped by the running `Rails.env`. That's why the code on `ConnectionSpecification::Resolver` didn't change much. Internally we still handle a 2 levels config Hash.
However, we are keeping the `configurations` getter/setter to be backward compatible and to be able to access it at the tasks level. (We might be able to deprecate this moving forward, that's another discussion tho)
The good thing about this, is that I was able to remove the `Rails.env` call from inside AR connection code. Now AR won't depend on the RAILS_ENV anymore.

### Caveats
If you were using a non-default `database.yml`, for instance:
```
development:
  adapter: sqlite3
development_readonly:
  adapter: sqlite3
```
There will be some necessary changes in the app, as AR will scope the config by the environment, so AR would only see:
```
development:
  adapter: sqlite3
```

The upgrade path would be to move the `development_readonly` config to inside `development` and reference it when calling `establish_connection`. (*see the third task on future work)

### Future work

- [ ] Change the guides.
- ~~Change the default database.yml template~~
- [ ] Connect to all databases under the env during Rails initializer, so apps won't need to call `establish_connection(:readonly)` for instance.
- [ ] Deprecate `create:all`/`drop:all` and make `create`/`drop` create all the dbs under the development/test envs?
- ~~Deprecate the old template/config with a warning message?~~

All the future work are not in this PR, because I would like to discuss one by one before implementing them, to make sure they make sense.

review @kaspth @dhh @rafaelfranca @tenderlove @jeremy 
@schneems (specially on commit https://github.com/rails/rails/commit/e51edd46a9cc1f4e0ab5ed6aac47c13a508940db, because it changes a lot from 6cc03675d30b58e28f585720dad14e947a57ff5b, However, I made sure the core functionality still works, which is make `DATABASE_URL` work without Rails.)

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 1]
```diff
@@ -17,6 +17,26 @@ def to_hash
         @config.merge(name: @name)
       end
 
+      # Converts a one level connection config, to the new format
+      # which is a two level, such as: {'primary' => {'adapter' => 'foo'}}
+      class LegacyConfigTransformer # :nodoc:
```
Comment (kirs) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: I don't think we should call it a legacy config. It's still a valid config that most of Rails apps are going to use, as long as we don't deprecate one level config.

**Comment by matzke** [comment]
+1 for decoupling Rails.env call from inside AR connection code

-1 for 3 level configuration as a "default", because for simple projects setup is imho way too complex. if 3 levels work that's fine, but "default" should be the simple version

**Comment by siegy22** [comment]
Why do you use `ENV.fetch("RAILS_MAX_THREADS") { 5 }` instead of `ENV.fetch("RAILS_MAX_THREADS", 5)`? :grin:

**Comment by dhh** [comment]
Like the new format, but concur with @matzke that most apps won't need multiple DBs starting out, so the default database.yml should still be a standard 2-level setup. Then we can explain the 3-level setup in the guide. So that informs that whatever we call the code isn't about "legacy", it's about 2-level depth or 3-level depth.

**Comment by arthurnn** [comment]
👍  thanks @matzke and @dhh for the comment on the legacy naming.
I, indeed, thought about it multiple times. I agree that 98% of the rails apps won't probably need the 3 level connections, so let's not call it legacy, and make it live in our code base as a proper 'transformer' stage.

**Review by kaspth — COMMENTED** [review]
Some initial rough thoughts and suggestions 😊

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 2]
```diff
@@ -864,7 +864,7 @@ def connection_pool_list
       alias :connection_pools :connection_pool_list
 
       def establish_connection(config)
-        resolver = ConnectionSpecification::Resolver.new(Base.configurations)
+        resolver = ConnectionSpecification::Resolver.new(Base.local_configurations)
```
Comment (maclover7) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb]: 👍 agree with Kasper, the naming here is difficult to understand; it took me several read throughs to understand the difference between `local_configurations` and `configurations` 😬

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 3]
```diff
@@ -126,42 +156,35 @@ def initialize(configurations)
...
-            config[key] = resolve(value) if value
-          end
-          config
-        end
-
         # Returns an instance of ConnectionSpecification for a given adapter.
-        # Accepts a hash one layer deep that contains all connection information.
+        # Accepts:
+        # - Hash: one layer deep Hash that contains all connection information
+        # - Symbol: a connection specification name, that will be looked-up
```
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: 👍  Agree @kaspth ,

So the model using a readonly config could look like this:

```ruby
class User < ApplicationRecord
  self.connection_configuration = :readonly
end
```

or:
```ruby
class User < ApplicationRecord
  def self.connection_configuration
    Application.readonly_mode? ? :readonly : :primary
  end
end
```

Pretty much we could replace the `connection_specification_name` to `connection_configuration` in the model.

I don't want to change those in this PR tho, we can try in a next one. Also there will be some deprecation cycle for the name change, as 5.0 uses `connection_specification_name`

**Review by arthurnn — COMMENTED** [review]
I clean-up this a bit. There is no legacy reference anymore, as we will support 2 level configs moving forward.
@kaspth I liked your suggestion on having a Configuration class, that's what I did.
However, I didn't want to extend Hash, and instead, I did a Proxy. (http://words.steveklabnik.com/beware-subclassing-ruby-core-classes)
Also, I cannot normalize the config in the initializer, as `AR::Base.configurations` is a public API, and people could be using that configuration for something. so instead I just created a `normalize` method, that will return the config the way the resolver expects.

Thoughts?

**Comment by arthurnn** [comment]
@matthewd as discussed. I changed the PR to make this possible:
database.yml
```
production:
  database: ...
production_readonly:
  database: ...
```

So, old setups, with a 2 levels config, will still work, and will still be able to do a `establish_connection(:production_readonly)`

**Comment by kirs** [comment]
Do we have any plan about how this works with migrations?

I see at least two scenarios:

1) You have `primary` and `readonly` connections and you want to use the first one to run the migrations.

2) You have different connections with different databases. In our case, `master` and `shard_X` databases have completely different tables. For every migration, we have a method that describes which database the migration is going to alter.

**Comment by kaspth** [comment]
@kirs I think `bin/rails g migration some_name --db readonly` with `self.connection = :readonly` in the migration would suffice, but maybe we should shelve that discussion for later. Think this PR has big implications as is 😊

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 4]
```diff
@@ -17,6 +17,37 @@ def to_hash
         @config.merge(name: @name)
       end
 
+      class ConnectionConfigurations < SimpleDelegator #:nodoc:
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Isn't inheriting from SimpleDelegator effectively the same thing as inheriting from Hash?

Maybe we should just restrict the allowed interface to `[]`?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 5]
```diff
@@ -17,6 +17,37 @@ def to_hash
...
+        attr_accessor :root_level
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Since we're only using the instance variable there's no reason for the reader method. I'd do: `attr_writer`.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 6]
```diff
@@ -17,6 +17,37 @@ def to_hash
...
+          @root_level = nil
+        end
+
+        def at(top_level)
+          normalized(top_level)
+        end
+
+        def normalized(key = @root_level)
+          config = key ? self[key] : self
+          config = config ? config.dup : {}
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Find these two lines hard to parse because of the two conditionals. The empty hash fallback seems too defensive at first sight. What case is it trying to cover? That there's a key in `self[key]` but with no config at all?

Could we simplify to this?

```ruby
config = fetch(key, self).dup
```

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 7]
```diff
@@ -17,6 +17,37 @@ def to_hash
...
+
+          # if the configuration is a one level config, pushes that to be a two level config, with primary as key
+          if !config.key?("primary") && config.key?("adapter")
+            config = { "primary" => config }
+          end
+
+          if url = ENV["DATABASE_URL"]
+            config["primary"] ||= {}
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: We've just spent lines establishing `"primary"`, surely it should have been made an empty hash by now?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 8]
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
...
-        #   Resolver.new(configurations).resolve(:production)
+        #   configurations = { "primary" => "postgresql://localhost/foo" }
+        #   Resolver.new(configurations).resolve(:primary)
         #   # => { "host" => "localhost", "database" => "foo", "adapter" => "postgresql" }
         #
         def resolve(config)
           if config
             resolve_connection config
-          elsif env = ActiveRecord::ConnectionHandling::RAILS_ENV.call
-            resolve_symbol_connection env.to_sym
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Are we removing this because `resolve_connection` calls it?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 9]
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
...
         end
 
-        # Expands each key in @configurations hash into fully resolved hash
-        def resolve_all
-          config = configurations.dup
-          config.each do |key, value|
-            config[key] = resolve(value) if value
-          end
-          config
-        end
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: This makes me wonder how much we need to worry about people calling all the available Hash methods on `ActiveRecord::Base.configurations[some_env]` and if we can just normalize the configurations up front.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 10]
```diff
@@ -223,19 +249,21 @@ def resolve_connection(spec)
             end
           end
 
-          # Takes the environment such as +:production+ or +:development+.
+          # Takes the specification name such as +:primary+.
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: `configuration name` 😉

**Review by kaspth — COMMENTED** [review]
Some more feedback ❤️

File: activerecord/lib/active_record/connection_handling.rb [hunk 11]
```diff
@@ -44,10 +41,9 @@ module ConnectionHandling
     #
     # The exceptions AdapterNotSpecified, AdapterNotFound and +ArgumentError+
     # may be returned on an error.
-    def establish_connection(config = nil)
+    def establish_connection(config = :primary)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_handling.rb]: I think `self.configuration` is way too vague to configure a model. In the bowels of AR, it's fine to infer that "configuration" means "Active Record configuration", and then assume that means "[database connection] configuration".

At the model layer, though, we should be more explicit: `User.configuration` is just unclear.

**Comment by sirupsen** [comment]
I like the new format as well, we have something like this supported by monkey-patches internally. I would share @kirs concerns about migrations. Does the schema cache do the right thing? That's an area where we'd be happy to get rid of our custom code.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 12]
```diff
@@ -126,42 +159,35 @@ def initialize(configurations)
...
+        # - Symbol: a configuration name that will be looked-up
+        #           from the configurations Hash
+        # - String: a database url
         #
         # == Example
         #
-        #   config = { "production" => { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" } }
-        #   spec = Resolver.new(config).spec(:production)
+        #   config = { "host" => "localhost", "database" => "foo", "adapter" => "sqlite3" }
+        #   spec = Resolver.new({}).spec(config)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: This API looks weird to me. I'd expect the resolver to always be initialized with the config and resolve a spec.

File: activerecord/lib/active_record/railties/databases.rake [hunk 13]
```diff
@@ -324,22 +324,22 @@ db_namespace = namespace :db do
       begin
         should_reconnect = ActiveRecord::Base.connection_pool.active_connection?
         ActiveRecord::Schema.verbose = false
-        ActiveRecord::Tasks::DatabaseTasks.load_schema ActiveRecord::Base.configurations["test"], :ruby, ENV["SCHEMA"]
+        ActiveRecord::Tasks::DatabaseTasks.load_schema ActiveRecord::Tasks::DatabaseTasks.config_at("test"), :ruby, ENV["SCHEMA"]
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/railties/databases.rake]: Looks like it indeed broke.

File: activerecord/test/cases/tasks/database_tasks_test.rb [hunk 14]
```diff
@@ -152,23 +156,42 @@ def test_creates_configurations_with_blank_hosts
...
   end
 
   class DatabaseTasksCreateCurrentTest < ActiveRecord::TestCase
     def setup
+      @prev_config = ActiveRecord::Base.configurations
       @configurations = {
-        "development" => { "database" => "dev-db" },
-        "test"        => { "database" => "test-db" },
-        "production"  => { "database" => "prod-db" }
+        "development" => { "database" => "dev-db", "adapter" => "sqlite3" },
```
Comment (rafaelfranca) [inline activerecord/test/cases/tasks/database_tasks_test.rb]: Why the adapter is now required?

Comment (kaspth) [inline activerecord/test/cases/tasks/database_tasks_test.rb]: I think it's because we only check for the `adapter` key at the moment. Finding a more accurate way to find an actual config could fix the need to change this.

File: activerecord/lib/active_record/connection_handling.rb (hunk 11 above)
Comment (kaspth) [inline activerecord/lib/active_record/connection_handling.rb]: Agreed, it's too broad. Perhaps I should just learn to stop worrying and love `self.connection_specification_name =` 😄

File: activerecord/test/cases/tasks/database_tasks_test.rb (hunk 14 above)
Comment (matthewd) [inline activerecord/test/cases/tasks/database_tasks_test.rb]: It's already required in practice for an actual config you intend to use, and while this test was accidentally assuming it, we've never claimed to be an arbitrary hash storage bucket.

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 8 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: added back

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb (hunk 12 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: fixed

File: activerecord/lib/active_record/railties/databases.rake (hunk 13 above)
Comment (arthurnn) [inline activerecord/lib/active_record/railties/databases.rake]: adding it back

**Comment by kaspth** [comment]
Looks like the backwardscompatibility issues have been solved, so I'd say rebase, fix tests, run `rubocop -a`, squash and then you have my ok to merge 👍

This was a hard and long haul so I definitely appreciate your willingness to run my review gauntlet so many times 😄

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 15]
```diff
@@ -17,6 +17,45 @@ def to_hash
...
+            ret = c[key]
+          end
+          ret ||= @config[key]
+          if @database_url_config
+            ret = (ret || {}).merge(@database_url_config)
+          end
+          ret
+        end
+
+        def keys
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: When is this called? My cmd-E on `keys` can't find any instances?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 16]
```diff
@@ -17,6 +17,45 @@ def to_hash
...
+            @config[@root_level].keys
+          else
+            @config.keys
+          end
+        end
+
+        def to_hash
+          @config
+        end
+      end
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: I still found a lot of this logic hard to read, so I took the liberty of trying to make it more concise:

```ruby
class ConnectionConfigurations #:nodoc:
  def initialize(config)
    @root_config = @config = config
    @database_url_config = resolve_database_url_config || {}
  end
 
  def scope=(scope)
    @root_config, @config = @config, (@config[scope] || {})
  end
 
  def [](key)
    @database_url_config.reverse_merge(@config[key] || @root_config[key] || {})
  end

  def keys
    @config.keys
  end

  def to_h
    @config
  end

  private
    def resolve_database_url_config
      if ENV["DATABASE_URL"]
        ConnectionUrlResolver.new(ENV["DATABASE_URL"]).to_hash
      end
    end
end
```

I'm curious why we need the `@root_config[key]` fallback in `[]`?

NOTE: `scope=` basically assumes that it's never assigned more than once, which it doesn't seem to be in library code, but it is in test code. Would it make more sense to pass it in at the constructor? Or have a `scope` method that returns a new instance of `ConnectionConfigurations`?

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 17]
```diff
@@ -17,6 +17,45 @@ def to_hash
...
+        attr_accessor :root_level
+
+        def [](key)
+          ret = nil
+          if @root_level && c = @config[@root_level]
+            ret = c[key]
+          end
+          ret ||= @config[key]
+          if @database_url_config
+            ret = (ret || {}).merge(@database_url_config)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: Isn't this applying the URL-supplied database to every configuration we return?

That's wrong, and I'm alarmed the tests aren't point that out. It's hard to see what's going on with them, because they're all renamed at the same time, though. 😟
